src/ngraph/frontend/onnx_import/op/matmul.cpp

```diff
@@ -0,0 +1,181 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include <algorithm>
+#include <cstddef>
+#include <functional>
+#include <iterator>
+#include <memory>
+#include <numeric>
+#include <vector>
+
+#include "ngraph/assertion.hpp"
+#include "ngraph/coordinate.hpp"
+#include "ngraph/op/concat.hpp"
+#include "ngraph/op/dot.hpp"
+#include "ngraph/op/reshape.hpp"
+#include "ngraph/op/slice.hpp"
+
+#include "exceptions.hpp"
+#include "matmul.hpp"
+#include "utils/broadcasting.hpp"
+#include "utils/common.hpp"
+#include "utils/reshape.hpp"
+
+/// \brief      Slice the sub matrix from 3D input tensor.
+///
+/// \param[in]  node  The input tensor. Must be 3D.
+/// \param[in]  idx   The index on the first axis, at wihich to slice sub-matrix.
+///
+/// \return     The node representing sub matrix.
+///
+static std::shared_ptr<ngraph::Node> get_sub_matrix(const std::shared_ptr<ngraph::Node>& node,
+                                                    std::size_t idx)
+{
+    ngraph::Shape shape{node->get_shape()};
+    ngraph::Coordinate lower_bounds(shape.size());
+    ngraph::Coordinate upper_bounds = shape;
+
+    lower_bounds.at(0) = idx;
+    upper_bounds.at(0) = idx + 1;
+
+    auto sub_matrix{std::make_shared<ngraph::op::Slice>(node, lower_bounds, upper_bounds)};
+    // Remove first single entry dim.
+    ngraph::Shape output_shape{std::next(std::begin(shape)), std::end(shape)};
+    return {std::make_shared<ngraph::op::Reshape>(
+        sub_matrix,
+        ngraph::onnx_import::reshape::get_default_axis_vector(sub_matrix->get_shape().size()),
+        output_shape)};
+}
+
+namespace ngraph
+{
+    namespace onnx_import
+    {
+        namespace op
+        {
+            namespace set_1
+            {
+                NodeVector matmul(const Node& node)
+                {
+                    NodeVector ng_inputs{node.get_ng_inputs()};
+                    auto left{ng_inputs.at(0)};
```

Comment:
A lot of copying is going on here - you should be able to use const references here too.